text/0811-element-modifiers.md

```diff
@@ -0,0 +1,232 @@
+
+function cleanup(instance: OnModifier) {
+  let { element, event, handler } = instance;
+
+  if (element && event && handler) {
+    element.removeEventListener(event, handler);
+
+    instance.element = null;
+    instance.event = null;
+    instance.handler = null;
+  }
+}
+
+export default class OnModifier extends Modifier {
+  element = null;
+  event = null;
+  handler = null;
+
+  // Manage teardown
+  constructor() {
+    super(...arguments);
+    registerDestructor(this, cleanup);
+  }
+
+  // Run on installation and all changes to tracked state.
+  modify(element, [event, handler]) {
+    // Clear any previous state.
+    cleanup(this);
+
+    // Set up the new handling.
+    this.addEventListener(element, event, handler);
+  }
+
+  // methods for reuse
+  addEventListener(element, event, handler) {
+    // Store the current element, event, and handler for when we need to remove
+    // them during cleanup.
+    this.element = element;
+    this.event = event;
+    this.handler = handler;
+
+    element.addEventListener(event, handler);
+  };
+}
+```
+
+While this is slightly more complicated than the function-based version,
+that complexity comes along with much more *control*. It also allows us to hook
+into Ember's dependency injection system, for example to use services.
+
+[RFC #432: Contextual Helpers and Modifiers][rfc-0432] allows optional
+invocation and partial application of modifiers, and allows modifiers to be
+yielded, passed as arguments, etc. [RFC #779: First-Class Component
+Templates][rfc-0779] allows defining them locally or importing them directly to
+be used in a `<template>` context. This combination makes many patterns much
+easier to implement:
+
+```js
+// /app/components/slide-up-card.js
+import Component from '@glimmer/component';
+import { modifier } from 'ember-modifier';
+
+const lockBodyScroll = modifier((element, [shouldLockBodyScroll]) => {
+  document.body.classList.add('lock-scroll');
+  return () => document.body.classList.remove('lock-scroll');
+});
+
+export default class SlideUpCard extends Component {
+  @action closeCard (event) {
+    event.preventDefault();
+    this.args.close();
+  }
+
+  <template>
+    {{#if has-block}}
+      {{yield (hash lock=lockBodyScroll onClick=this.closeCard)}}
+    {{else}}
+      <div
+        class="slide-up-card"
+        ...attributes
+        {{lockBodyScroll}}
+      >
+        <p>{{@bodyContent}}</p>
+
+        <button {{on "click" this.closeCard}}>Dismiss</button>
+      </div>
+    {{/if}}
+  </template>
+}
+```
+
+## Detailed design
+
+The necessary changes to `ember-cli` are relatively small since we only need
+to add the dependency to the `app` blueprint, and the `addon` blueprint will
+inherit it automatically.
+
+This has the advantage (over including it as an implicit dependency), that
+apps and addons that don't want to use it for some reason can opt out by
+removing the dependency from their `package.json` file.
+
```

Comment:
We should add a comment here explicitly calling out that "add an ecosystem addon to the blueprint" is not our normal operating model. Instead, it is a case where we have carefully rationalized that design *with an eye to doing this*, and we *know* it is introducing additional "incoherence" to the programming model.

```suggestion

**Notes:**

1.  This is *not* the usual path for delivering features into Ember: we do not
    generally introduce community addons directly into the blueprint. In this
    specific case, we think it is the right move:
        
        -   The addon has intentionally been reworked to rationalize it in
            terms of the rest of the Octane programming model, explicitly with an eye to adoption in this way.

        -   We do *not* want to introduce `@glimmer/modifier` with exactly this
            API, at least not yet, because we believe we may want to introduce
            a slightly updated API from that package *without* requiring
            breaking changes in the future.

2.  We are intentionally introducing more "incoherence" to the programming
    model with this addition. (Or rather, we are acknowledging the *existing*
    incoherence in the ecosystem, with an eye to making progress on resolving
    it!) In particular, as the community begins adopting strict mode templates
    via First-Class Component Templates in the months ahead, they will often
    end up importing from both `ember-modifier` and `@ember/modifier`:

    ```js
    import { on } from '@ember/modifier';
    import { modifier } from 'ember-modifier';

    const playWhen = modifier((el, [shouldPlay]) => {
      if (shouldPlay) {
        el.play();
      } else {
        el.pause();
      }
    });

    <template>
      <audio
        src={{@src}}
        {{on "error" @onBadLoad}}
        {{playWhen @shouldPlay}}
      />
    </template>
    ```

    As noted in (1) above, we have ideas on how to resolve this, but are
    intentionally not blocking on those in favor of unlocking this key
    functionality for Ember users in the Octane programming model *today*.

```

Reply:
@chriskrycho thank you! these notes were added to the RFC